packages/cohort/src/replicator/services/statemap_installer_service.rs

```diff
@@ -1,35 +1,80 @@
+    statemap_installer: Arc<dyn ReplicatorInstaller + Send + Sync>,
+    mut installation_rx: mpsc::Receiver<(u64, Vec<StatemapItem>)>,
+    statemap_installation_tx: mpsc::Sender<StatemapInstallationStatus>,
+    config: StatemapInstallerConfig,
+) -> Result<(), String> {
+    // TODO: Pass the number of permits over an environment variable?
+    let permit_count = config.parallel_thread_count.unwrap_or(50) as usize;
+    let semaphore = Arc::new(Semaphore::new(permit_count));
 
     loop {
-        if let Some(batch) = statemaps_rx.recv().await {
-            for (ver, statemap_batch) in batch {
-                debug!("[Statemap Installer Service] Received statemap batch ={statemap_batch:?} and version={ver:?}");
-                match statemap_installer.install(statemap_batch, Some(ver)).await {
-                    Ok(true) => {
-                        replicator_tx.send(ReplicatorChannel::InstallationSuccess(vec![ver])).await.unwrap();
-                    }
-                    Ok(false) => {
-                        // Do nothing if result is false.
-                    }
-                    Err(err) => return Err(err.to_string()),
-                }
-            }
-        }
+        let available_permits = semaphore.available_permits();
+
+        if available_permits > 0 {
+            if let Some((ver, statemaps)) = installation_rx.recv().await {
+                let replicator_tx_clone = replicator_tx.clone();
+                let statemap_installation_tx_clone = statemap_installation_tx.clone();
+                let installer = Arc::clone(&statemap_installer);
+
+                let permit = semaphore.clone().acquire_owned().await.unwrap();
+                tokio::spawn(async move {
+                    debug!("[Statemap Installer Service] Received statemap batch ={statemaps:?} and version={ver:?}");
+                    let start_installation_time = Instant::now();
+
+                    match installer.install(statemaps, Some(ver)).await {
+                        Ok(status) => {
+                            // let end_installation_time = start_installation_time.elapsed();
+                            // error!("[installation_service] Installed successfully version={ver} in {end_installation_time:?}");
+                            replicator_tx_clone.send(ReplicatorChannel::InstallationSuccess(vec![ver])).await.unwrap();
+                            match status {
+                                ReplicatorInstallStatus::Success => {
+                                    statemap_installation_tx_clone.send(StatemapInstallationStatus::Success(ver)).await.unwrap();
+                                }
+                                ReplicatorInstallStatus::Gaveup(_) => {
+                                    statemap_installation_tx_clone.send(StatemapInstallationStatus::GaveUp(ver)).await.unwrap();
+                                }
+                            }
+
+                            drop(permit);
```

Comment:
Shouldn't we wait on spawn join handle and then drop permit. This will keep permit lifespan aligned with "thread" lifespan.